### PR TITLE
Use sabre for vcard parsing

### LINKED ISSUE #128: Use Vcard library
It seems http://sabre.io/vobject/vcard/ might be a better choice than our embeded library. Need to retest if extended attributes are possible.

### PR DESCRIPTION
**Pull request by blacksenator**
complete conversion so that sabre/vobject fits

Replaces #130, fixes #128.

### CODE REVIEW

File: src/CardDav/Backend.php [hunk 1]
```diff
@@ -171,61 +167,43 @@ public function getVcards(): array
...
-                $types = "TYPE=" . strtoupper($embedded['subtype']) . ";ENCODING=b";
-                break;
-            case '4.0':
-                $types = "data:" . $embedded['mimetype'] . ";base64";
-                break;
+        if ($embedded = $this->getLinkedData($vcard->$property)) {      // get the data from the external URL or false
+            if ($vcard->VERSION == '3.0') {                             // the different vCard versions must be considered
+                unset($vcard->$property);                               // delete the old property
+                $vcard->add($property, $embedded['data'], ['TYPE' => strtoupper($embedded['subtype']), 'ENCODING' => 'b']);
+            } elseif ($vcard->VERSION == '4.0') {
```
Comment (blacksenator) [inline src/CardDav/Backend.php]: see [#458](https://github.com/sabre-io/vobject/issues/458)

File: src/functions.php [hunk 2]
```diff
@@ -100,41 +98,63 @@ function uploadImages(array $vcards, array $config, array $phonebook, callable $
...
+            continue;
+        }
+        // Fritz!Box only accept jpg-files
+        if ($vcard->VERSION == '3.0') {
+            if ($vcard->PHOTO['TYPE'] != 'JPEG') {
+                continue;
+            } else {
+                $vcardImage = $vcard->PHOTO;
+            }
+        } elseif ($vcard->VERSION == '4.0') {                       // see: https://github.com/sabre-io/vobject/issues/458
```
Comment (blacksenator) [inline src/functions.php]: see [#458](https://github.com/sabre-io/vobject/issues/458)

File: src/CardDav/Backend.php [hunk 3]
```diff
@@ -238,90 +216,62 @@ public function getLinkedData(string $uri): array
...
-
-        return $response;
+        @list(
+            $lastname,
+            $firstname,
+            $additional,
+            $prefix,
+            $suffix
+        ) = explode(';', $value);
+        return (object) [
```
Comment (andig) [inline src/CardDav/Backend.php]: not sure this cast is valid in php 7.1?

File: tests/ConverterTest.php [hunk 4]
```diff
@@ -32,12 +33,12 @@ private function defaultConfig(): array
         ];
     }
 
-    private function defaultContact(): stdClass
+    private function defaultContact(): object
```
Comment (andig) [inline tests/ConverterTest.php]: object type hint was only introduced with php 7.2.

Also- for most other functions- this is omitted as well.

**Comment by andig** [comment]
Unfortunately the PR has conflicts. Would it be possible to rebase onto current master? Do this by pulling master onto master, switch to this branch and run `git rebase master`.

Be sure to create a copy of your branch (by branching again) before doing this in case anything goes wrong.

**Comment by andig** [comment]
Kein Rebase- die Konflikte sind weiter da. Master vorher lokal aktualisiert?

**Comment by andig** [comment]
Looks like more problems with object typehints?

File: src/functions.php [hunk 5]
```diff
@@ -121,41 +120,66 @@ function uploadImages(array $vcards, array $config, array $phonebook, callable $
...
-                    // upload of new image done, now store new image URL in vCard (new Random Postfix!)
-                    $vcard->imageURL = $imgPath . $newFTPimage;
-                } else {
-                    error_log(PHP_EOL."Error uploading $newFTPimage.");
-                    unset($vcard->rawPhoto);                           // no wrong link will set in phonebook
-                    unset($vcard->imageURL);                           // no wrong link will set in phonebook
-                }
-                fclose($memstream);
+        $countAllImages++;
+        $imgURL = 'IMAGEURL';                                       // usualy $vcard->IMAGEURL is sufficient - but Travis CI disagrees
```
Comment (andig) [inline src/functions.php]: This is not a valid fix and not needed. Please undo this change.

File: src/functions.php [hunk 6]
```diff
@@ -173,23 +197,22 @@ function uploadImages(array $vcards, array $config, array $phonebook, callable $
...
-            if (array_key_exists($vcard->fullname, $groups)) {
-                $groups[$vcard->fullname] = array_merge($groups[$vcard->fullname], $vcard->xabsmember);
-            } else {
-                $groups[$vcard->fullname] = $vcard->xabsmember;
+        if (isset($vcard->{'X-ADDRESSBOOKSERVER-KIND'})) {
+            if ($vcard->{'X-ADDRESSBOOKSERVER-KIND'} == 'group') {      // identifier
+                foreach ($vcard->{'X-ADDRESSBOOKSERVER-MEMBER'} as $member) {
+                    $groups[(string)$vcard->FN][] = (string)$member;
+                }
+            unset($vcards[$key]);                                       // delete this vCard
```
Comment (andig) [inline src/functions.php]: Wrong indent.

File: src/functions.php [hunk 7]
```diff
@@ -309,18 +333,8 @@ function filterMatches($attribute, $filterValues): bool
...
-                if ($childAttribute === $filter) {
-                    return true;
-                }
-            }
-        } else {
-            // check if simple attribute matches
-            if ($attribute === $filter) {
-                return true;
-            }
+        if (strpos($attribute, $filter) !== false) {
```
Comment (andig) [inline src/functions.php]: Is there a test for this change?

Comment (blacksenator) [reply src/functions.php]: Only paractic tests with real data back then. To be honest, this change is so long ago that I can not remember what made me shorten it. It took quite a long time for me to understand your coding. If I look at it now, I'm back in the beginning

Comment (andig) [reply src/functions.php]: The original change allowed both arrays and strings as attributes. If there is no reason to not do or need this and no covering test case I'd propose to not change it.

File: src/FritzBox/Converter.php [hunk 8]
```diff
@@ -132,79 +149,72 @@ private function addEmail(array $addresses)
...
+        foreach ($card->TEL as $key => $number) {
+            // format number
+            if (count($replaceCharacters)) {
+                $number = str_replace("\xc2\xa0", "\x20", $number);
+                $number = strtr($number, $replaceCharacters);
+                $number = trim(preg_replace('/\s+/', ' ', $number));
+            }
+            // get type
+            $type = 'other';
+            $telTypes = isset($card->TEL[$key]->parameters['TYPE']) ? strtoupper($card->TEL[$key]->parameters['TYPE']) : '';
```
Comment (andig) [inline src/FritzBox/Converter.php]: You can use the ?: operator to simplify this.

File: src/FritzBox/Converter.php [hunk 9]
```diff
@@ -215,57 +225,56 @@ private function getPhoneNumbers(stdClass $card): array
...
-                        $addAddress['classifier'] = $value;
-                        break;
-                    }
+        foreach ($card->EMAIL as $key => $address) {
+            $addAddress = [
+                'id'    => count($mailAdresses),
+                'email' => (string)$address,
+            ];
+
+            $mailTypes = isset($card->EMAIL[$key]->parameters['TYPE']) ? strtoupper($card->EMAIL[$key]->parameters['TYPE']) : '';
```
Comment (andig) [inline src/FritzBox/Converter.php]: Use ?: to simplify

File: src/functions.php [hunk 10]
```diff
@@ -83,16 +81,17 @@ function getFtpConnection($url, $user, $password, $directory, $secure)
 /**
  * upload image files via ftp to the fritzbox fonpix directory
  *
- * @param stdClass[] $vcards downloaded vCards
+ * @param object $vcards downloaded vCards
```
Comment (andig) [inline src/functions.php]: Should be `mixed[]` or array?

File: src/functions.php [hunk 11]
```diff
@@ -83,16 +81,17 @@ function getFtpConnection($url, $user, $password, $directory, $secure)
...
  * @param array $config
  * @param array $phonebook
  * @param callable $callback
  * @return mixed false or [number of uploaded images, number of total found images]
  */
-function uploadImages(array $vcards, array $config, array $phonebook, callable $callback=null)
```
Comment (andig) [inline src/functions.php]: Why remove the array type hint?

File: src/functions.php [hunk 12]
```diff
@@ -83,16 +81,17 @@ function getFtpConnection($url, $user, $password, $directory, $secure)
...
+function uploadImages($vcards, array $config, array $phonebook, callable $callback=null)
```
Comment (andig) [inline src/functions.php]: Missing array typehint for $vcards?

File: src/functions.php (hunk 7 above)
Comment (blacksenator) [reply src/functions.php]: I walked again through the whole filter functions again.
I found a few bugs (in dissolveGroups() and config.example.php-> filters), deviations (in filter ()) and optimizations (in filtersMatch()). The latter makes filterMatches() superfluous.
Thus the resulting commit contains a few more changes!

Since we do not yet have a testcase for the filter function, it will take some time until I get it correctly defined on top :(

File: src/FritzBox/Converter.php (hunk 8 above)
Comment (blacksenator) [reply src/FritzBox/Converter.php]: cool - the [Elvis operator](https://stackoverflow.com/questions/1993409/operator-the-elvis-operator-in-php) :)
But I´m checking against isset() and need to get strtoupper() - so I´m a little bit puzzled how this fits into the shortend syntax

File: src/FritzBox/Converter.php (hunk 9 above)
Comment (blacksenator) [reply src/FritzBox/Converter.php]: dito
this way?:
$mailTypes = strtouppper(isset($card->EMAIL[$key]->parameters['TYPE'])) ?: '';

File: src/FritzBox/Converter.php (hunk 8 above)
Comment (andig) [reply src/FritzBox/Converter.php]: Elvis? Nice. You can use Elvis inside and wrap then:

    $types = strtoupper(... ?: „“)

A little less efficient but more readable.

File: src/FritzBox/Converter.php (hunk 9 above)
Comment (andig) [reply src/FritzBox/Converter.php]: Almost- last brackts need to move before ;

File: src/FritzBox/Converter.php (hunk 8 above)
Comment (blacksenator) [reply src/FritzBox/Converter.php]: > Good idea - wrong method
(Douglas Adams - Last chance to see)

First time I consent your tests: change to
`$telTypes = strtoupper(isset($card->TEL[$key]->parameters['TYPE']) ?: '');`
results with 3 errors

File: src/CardDav/Backend.php [hunk 13]
```diff
@@ -251,40 +233,40 @@ public function getLinkedData(string $uri): array
...
+    public function enrichVcard($vcard)
     {
-        $parser = new Parser($body);
-        $vcard = $parser->getCardAtIndex(0);
-
-        return $vcard;
-    }
+        /** @var \stdClass $vcard */
+        if (isset($vcard->FN)) {                                // redundant for downward compatibility
+            $property = 'FULLNAME';
```
Comment (blacksenator) [inline src/CardDav/Backend.php]: I know this is stupid, but the only way to shup Travis CI up.
I tried
`$vcard->FULLNAME = (string)$vcard->FN;`
and
`$vcard->add('FULLNAME', (string)$vcard->FN);`
which are both correct - but Travis disagrees!

File: config.example.php [hunk 14]
```diff
@@ -44,17 +44,18 @@
...
-            // if empty include all by default
+        'include' => [                                          // if empty include all by default
+            /*
+            'categories' => [],
+            'groups' => [],
+            */
         ],
 
         'exclude' => [
-            'category' => [
```
Comment (blacksenator) [inline config.example.php]: This was a bug anyway. It should always have been 'categories'

File: src/FritzBox/Converter.php [hunk 15]
```diff
@@ -171,7 +171,7 @@ private function getPhoneNumbers($card): array
             }
             // get type
             $type = 'other';
-            $telTypes = isset($card->TEL[$key]->parameters['TYPE']) ? strtoupper($card->TEL[$key]->parameters['TYPE']) : '';
+            $telTypes = strtoupper(isset($card->TEL[$key]->parameters['TYPE'])) ?: '';
```
Comment (andig) [inline src/FritzBox/Converter.php]: Das kann so nicht funktionieren! Das isset muss weg, Klammern nach hinten. Getestet.

File: src/FritzBox/Converter.php [hunk 16]
```diff
@@ -250,7 +250,7 @@ private function getEmailAdresses($card): array
                 'email' => (string)$address,
             ];
 
-            $mailTypes = isset($card->EMAIL[$key]->parameters['TYPE']) ? strtoupper($card->EMAIL[$key]->parameters['TYPE']) : '';
+            $mailTypes = strtoupper(isset($card->EMAIL[$key]->parameters['TYPE'])) ?: '';
```
Comment (andig) [inline src/FritzBox/Converter.php]: Dito

File: src/functions.php [hunk 17]
```diff
@@ -205,28 +207,32 @@ function dissolveGroups(array $vcards): array
...
-    $vcards = array_values($vcards);
+    $vcards = array_values($vcards);   // no clue, what this line should do
 
     // assign group memberships
     foreach ($vcards as $vcard) {
         foreach ($groups as $group => $members) {
             if (in_array((string)$vcard->UID, $members)) {
-                if (!isset($vcard->GROUP)) {
-                    $vcard->add('GROUP', []);
+                if (isset($vcard->GROUPS)) {
```
Comment (andig) [inline src/functions.php]: Groups statt group? Testcase?

File: src/functions.php [hunk 18]
```diff
@@ -243,12 +249,17 @@ function dissolveGroups(array $vcards): array
  */
 function filter(array $vcards, array $filters): array
 {
+    $vcf = '';
+    foreach ($vcards as $vcard) {
+        $vcf = $vcf . $vcard->serialize();
+    }
+    file_put_contents('TEST.txt', $vcf);
```
Comment (andig) [inline src/functions.php]: Überbleibsel?

File: src/functions.php [hunk 19]
```diff
@@ -243,12 +249,17 @@ function dissolveGroups(array $vcards): array
...
+
     // include selected
     $includeFilter = $filters['include'] ?? [];
 
-    if (countFilters($includeFilter)) {
+    if (countFilters($includeFilter) > 0) {
```
Comment (andig) [inline src/functions.php]: In dem Fall entspricht >0 dem true

File: config.example.php [hunk 20]
```diff
@@ -44,17 +44,18 @@
     ],
 
     'filters' => [
-        'include' => [
-            // if empty include all by default
+        'include' => [                                          // if empty include all by default
```
Comment (blacksenator) [inline config.example.php]: From my point of view easier to understand and to avoid [such discussions](https://www.ip-phone-forum.de/threads/carddav-to-fritzbox-2-0-neue-version-verf%C3%BCgbar.298117/post-2334259).

File: src/CardDav/Backend.php [hunk 21]
```diff
@@ -3,14 +3,14 @@
...
-use Andig\Vcard\Parser;
+use Sabre\VObject\Reader;
 use GuzzleHttp\Client;
 use GuzzleHttp\Exception\RequestException;
-use \stdClass;
 
 /**
  * @author Christian Putzke <christian.putzke@graviox.de>
  * @author Andreas Goetz <cpuidle@gmx.de>
+ * @author Volker Püschel <knuffy@anasco.de>
```
Comment (blacksenator) [inline src/CardDav/Backend.php]: I've earned it

File: src/functions.php [hunk 22]
```diff
@@ -173,37 +199,40 @@ function uploadImages(array $vcards, array $config, array $phonebook, callable $
...
+                }
+                unset($vcards[$key]);                                   // delete this vCard
             }
-            unset($vcards[$key]);
-            continue;
         }
     }
 
-    $vcards = array_values($vcards);
+    $vcards = array_values($vcards);   // no clue, what this line should do
```
Comment (blacksenator) [inline src/functions.php]: From my point of view, the line is superfluous

File: src/functions.php [hunk 23]
```diff
@@ -173,37 +199,40 @@ function uploadImages(array $vcards, array $config, array $phonebook, callable $
...
     foreach ($vcards as $key => $vcard) {
-        if (isset($vcard->xabsmember)) {
-            if (array_key_exists($vcard->fullname, $groups)) {
-                $groups[$vcard->fullname] = array_merge($groups[$vcard->fullname], $vcard->xabsmember);
-            } else {
-                $groups[$vcard->fullname] = $vcard->xabsmember;
+        if (isset($vcard->{'X-ADDRESSBOOKSERVER-KIND'})) {
+            if ($vcard->{'X-ADDRESSBOOKSERVER-KIND'} == 'group') {      // identifier
+                foreach ($vcard->{'X-ADDRESSBOOKSERVER-MEMBER'} as $member) {
+                    $member = str_replace(['urn:', 'uuid:'], ['', ''], (string)$member);
```
Comment (blacksenator) [inline src/functions.php]: iCloud use both prefix - but I´m not sure if this is fix - so I decided to replace them separatly

File: src/functions.php [hunk 24]
```diff
@@ -214,18 +243,23 @@ function dissolveGroups(array $vcards): array
...
- * @return stdClass[]
+ * @return mixed[]
  */
 function filter(array $vcards, array $filters): array
 {
+    $vcf = '';
+    foreach ($vcards as $vcard) {
+        $vcf = $vcf . $vcard->serialize();
+    }
+    file_put_contents('TEST.txt', $vcf);
```
Comment (blacksenator) [inline src/functions.php]: shit! debugging remains -> next commit :(

File: src/functions.php [hunk 25]
```diff
@@ -214,18 +243,23 @@ function dissolveGroups(array $vcards): array
...
+
     // include selected
     $includeFilter = $filters['include'] ?? [];
 
-    if (countFilters($includeFilter)) {
+    if (countFilters($includeFilter) > 0) {
```
Comment (blacksenator) [inline src/functions.php]: This was a bug

File: src/functions.php [hunk 26]
```diff
@@ -276,49 +310,18 @@ function countFilters(array $filters): int
...
-                if ($childAttribute === $filter) {
-                    return true;
-                }
-            }
-        } else {
-            // check if simple attribute matches
-            if ($attribute === $filter) {
+        $param = strtoupper($attribute);
+        if (isset($vcard->$param)) {
+            if (array_intersect($vcard->$param->getParts(), $values)) {
```
Comment (blacksenator) [inline src/functions.php]: Short and easy: getParts() delivers always an array and $values are an array depending on example.config.php.
Makes filterMatches superflous.

File: src/functions.php (hunk 17 above)
Comment (blacksenator) [reply src/functions.php]: yes - let´s have a look in FunctionsTest

File: src/FritzBox/Converter.php (hunk 15 above)
Comment (blacksenator) [reply src/FritzBox/Converter.php]: NO - neither way!
TestCase Okay! But...
Real life test:
Notice: Undefined index: TYPE in C:\Users\Admin\Documents\Development\Git\carddav2fb\src\FritzBox\Converter.php on line 174
Contact without phone numbers will be skipped
Because "Das isset muss weg,"

File: src/functions.php [hunk 27]
```diff
@@ -569,6 +561,8 @@ function getQuickdials(SimpleXMLElement $xmlPhonebook)
             }
         }
     }
+    ksort($quickdialNames);                                 // ascending: lowest quickdial first
```
Comment (blacksenator) [inline src/functions.php]: see following comment

File: config.example.php (hunk 20 above)
Comment (andig) [reply config.example.php]: Welches renaming? Ich versteh den Kommentar nicht. Willst du hier was ändern?

File: src/functions.php (hunk 25 above)
Comment (andig) [reply src/functions.php]: Ich versteh nicht? Ohne >0 gings nicht? Kann ich mir nicht vorstellen.

File: src/functions.php (hunk 27 above)
Comment (andig) [reply src/functions.php]: Versteh ich nciht. Was kommentierst du hier und warum? Dazu hab ich nichts gefragt?

File: src/CardDav/Backend.php (hunk 1 above)
Comment (andig) [reply src/CardDav/Backend.php]: 457 kann damit entfallen.

File: src/functions.php (hunk 2 above)
Comment (andig) [reply src/functions.php]: Ebenso?

File: src/CardDav/Backend.php (hunk 13 above)
Comment (andig) [reply src/CardDav/Backend.php]: Dennoch geht das so nicht. Wir verbasteln nicht den Code um Travis glücklich zu machen.

File: src/CardDav/Backend.php (hunk 3 above)
Comment (andig) [reply src/CardDav/Backend.php]: pls fix

File: tests/ConverterTest.php (hunk 4 above)
Comment (andig) [reply tests/ConverterTest.php]: pls fix

File: src/CardDav/Backend.php [hunk 28]
```diff
@@ -307,4 +288,25 @@ private function stripNamespaces($xml)
 
         return $xml;
     }
+
+    /**
+    * split the values from 'N' into separate fields
```
Comment (andig) [inline src/CardDav/Backend.php]: indent

File: src/FritzBox/Converter.php (hunk 8 above)
Comment (andig) [reply src/FritzBox/Converter.php]: So schwer ist es doch nicht wenn Du Dir anschaust was ?: tut. Das isset ist doch schon eingebaut!

    $telTypes = strtoupper($card->TEL[$key]->parameters['TYPE'] ?: '');

File: src/FritzBox/Converter.php (hunk 9 above)
Comment (andig) [reply src/FritzBox/Converter.php]: dito

File: src/functions.php (hunk 10 above)
Comment (andig) [reply src/functions.php]: ping

File: src/functions.php (hunk 11 above)
Comment (andig) [reply src/functions.php]: ping

File: src/functions.php (hunk 12 above)
Comment (andig) [reply src/functions.php]: ping

File: src/functions.php [hunk 29]
```diff
@@ -121,41 +120,66 @@ function uploadImages(array $vcards, array $config, array $phonebook, callable $
...
+                continue;
+            } else {
+                $vcardImage = (string)$vcard->PHOTO;
+            }
+        } elseif ($version == '4.0') {                       // see: https://github.com/sabre-io/vobject/issues/458
+            $value = explode(',', (string)$vcard->PHOTO, 2);
+            if (!preg_match("/jpeg/", $value[0])) {
+                continue;
+            } else {
+                if (!$vcardImage = base64_decode($value[1])) {      // PONR: I don´t trust this way of fetching data
```
Comment (andig) [inline src/functions.php]: was heisst das? hier wird nix gefetcht?

File: src/functions.php (hunk 5 above)
Comment (andig) [reply src/functions.php]: bitte beheben

File: src/FritzBox/Converter.php (hunk 15 above)
Comment (andig) [reply src/FritzBox/Converter.php]: siehe andere kommentare. isset muss weg.

**Comment by andig** [comment]
Der cs fixer bemängelt noch eine Leerzeile die gem. Styleguide nicht da sein sollte: https://travis-ci.org/andig/carddav2fb/jobs/570347198#L603

File: src/functions.php (hunk 29 above)
Comment (blacksenator) [reply src/functions.php]: Okay, missverständlich! Ich wollte damit ausdrücken, dass es seitens sabre für vCards der VERSION 4 keinen sauberen Type-/Parameter-basierten Zugriff gibt wie für VERSION 3 (siehe: "[Working with parameters](http://sabre.io/vobject/vcard/)").
Ausgangslage ist:
RFC 6350: 4.0: PHOTO:data:image/jpeg;base64,[base64-data]
Ein Zugriff anlaog dem Beispiel:
```
$type = $vcard->TEL['TYPE'];
echo (string)$type;
```
funktioniert mit 
`$mimetype = $vcard->PHOTO['data'];`
nicht. Alles was ich hinbekommen habe, ist den gesamten Inhalt zu nehmen und auseinander zu pflücken (explode) und dann auf die hoffentlich kosistenten Teile zuzugreifen (nicht sehr fehlertolerant).

File: src/functions.php (hunk 25 above)
Comment (blacksenator) [reply src/functions.php]: counterfilters liefert immer true! Auch wenn es null - also keine Definitionen enthält! Dann ist es aber nicht unsinnig die Bedingung zu durchlaufen, weil ja (keine Includedefinition = alles includen). Die IF bedingung (dezidierte Prüfung) zu durchlaufen ist nur notwendig wenn >= 1 Filterdefinition vorliegt.

File: src/functions.php (hunk 19 above)
Comment (blacksenator) [reply src/functions.php]: siehe unten

BTW
`countFilters()` liefert `int` und nicht `bool` zurück. Also immer 0 oder >0
Dann sollte die Abfrage m.E. auch darauf sich beziehen, weil ein einfaches
`if (countFilters($var))` (wie verwendet immer true zurückliefert.

File: src/FritzBox/Converter.php (hunk 15 above)
Comment (blacksenator) [reply src/FritzBox/Converter.php]: Wie ich geschrieben habe: wenn isset() weg ist, dannläuft zwar des TestConverter fehlerfrei durch (weil kein defaultContact definiert ist, der ohne Rufnummer ist!
Im Life-Test führt das zu:

> Notice: Undefined index: TYPE in C:\Users\Admin\Documents\Development\Git\carddav2fb\src\FritzBox\Converter.php

File: src/CardDav/Backend.php (hunk 13 above)
Comment (blacksenator) [reply src/CardDav/Backend.php]: built back to
`$vcard->FULLNAME = (string)$vcard->FN;`


as I said: Travis CI is a bastard - I really hate it :)

File: src/functions.php (hunk 25 above)
Comment (andig) [reply src/functions.php]: Siehe https://3v4l.org/dHe3B das ist NICHT nötig. Ich wiederhole mich da. Du behauptest es ist nötig. NEIN.

File: src/functions.php (hunk 12 above)
Comment (blacksenator) [reply src/functions.php]: build back a few commits ago -> outdated

File: src/functions.php (hunk 11 above)
Comment (blacksenator) [reply src/functions.php]: build back a few commits ago -> outdated

File: src/functions.php (hunk 10 above)
Comment (blacksenator) [reply src/functions.php]: build back a few commits ago -> outdated

File: src/FritzBox/Converter.php (hunk 9 above)
Comment (blacksenator) [reply src/FritzBox/Converter.php]: see below: $telTypes = ...

File: src/FritzBox/Converter.php (hunk 8 above)
Comment (blacksenator) [reply src/FritzBox/Converter.php]: und warum geht dann das script auf die Bretter (na ja NOTICE:), wenn ich Kontakte ohne Rufnummer durch den Konverter jage?

**Comment by blacksenator** [comment]
> 
> 
> Der cs fixer bemängelt noch eine Leerzeile die gem. Styleguide nicht da sein sollte: https://travis-ci.org/andig/carddav2fb/jobs/570347198#L603

Ja, ich weiß - aber ich finde sie um´s verrecken nicht! Any suggestions?

File: src/FritzBox/Converter.php (hunk 8 above)
Comment (andig) [reply src/FritzBox/Converter.php]: Schwer zu sagen ohne den Code mit Testcase zu sehen. Dafür sind die da.

File: src/CardDav/Backend.php [hunk 30]
```diff
@@ -307,4 +288,25 @@ private function stripNamespaces($xml)
 
         return $xml;
     }
+
+    /**
+     * split the values from 'N' into separate fields
+    */
```
Comment (andig) [inline src/CardDav/Backend.php]: Indent immer nich komisch. Kommentar für parameter und return value array?

File: src/functions.php (hunk 22 above)
Comment (andig) [reply src/functions.php]: Schmeiss raus, seltsam.

File: src/functions.php [hunk 31]
```diff
@@ -225,16 +254,15 @@ function filter(array $vcards, array $filters): array
...
         $step1 = [];
-
         foreach ($vcards as $vcard) {
             if (filtersMatch($vcard, $includeFilter)) {
                 $step1[] = $vcard;
             }
         }
     } else {
-        // filter defined but empty sub-rules?
+        // f
```
Comment (andig) [inline src/functions.php]: Typo?

File: src/functions.php [hunk 32]
```diff
@@ -586,6 +579,11 @@ function uploadBackgroundImage($phonebook, array $config)
         error_log('No quickdial numbers are set for a background image upload');
         return;
     }
+    if (key($quickdials) > 9) {    // usual the pointer should on the first element; with 7.3.*: array_key_first()
```
Comment (andig) [inline src/functions.php]: Wieso key? Ich denke das hängt vom array zustand ab und ist nicht stabil. Meinst du count? Oder den letzten eintrag? Oder den eintrag mit dem höchsten Wert?

Comment (blacksenator) [reply src/functions.php]: $key = quickdialnumber (is unique)
$value = Name

**Comment by andig** [comment]
Great job!!! See #150 to continue the discussion.
